### ISSUE TITLE
Feature request: Allow staging or unstaging of a single line or the selected lines

### ISSUE THREAD

**Issue by nelsonroliveira** [gitahead/gitahead#344]
# Request

Currently it's only possible to stage whole files or hunks but sometimes it's useful to allow for developers to commit only a few lines of a hunk.

## Use cases

* When developers start to make several unrelated changes at the same time, this makes it easier to make a commit that relates a single problem
* Some IDE's like to auto format/indent code and can make committing a mess if one is not careful to select only what was an intended change

## Examples on other products

![image](https://user-images.githubusercontent.com/10381502/70849352-11641500-1e75-11ea-88bc-db8f6e63b47e.png)

![image](https://user-images.githubusercontent.com/10381502/70849407-c565a000-1e75-11ea-9972-cf98ce98f822.png)

**Comment by dcendents** [comment]
I'd also love that feature. That's my biggest complain with GitAhead.

Maybe the way I work is not optimal, but I do TDD and I often have commented code lying around, which I do not want to commit yet, but is there because while I work on a test and the implementation, I already think of the next thing to test, so it's more of a personal reminder of the next things to do.

So right now I always need to keep an instance of git gui open just to be able to easily add individual lines and keep my comments out.

**Comment by KaKi87** [comment]
Hello,

Any news on this ?

Thanks !

**Comment by FinlayDaG33k** [comment]
I also would like this, right now I have to go back into my edittor, revert changes on an entire block just to commit that single line (or else the commit would make no sense).

**Comment by Murmele** [comment]
#401
In this pull request you can see my first version. Can you test it, so I can improve it and maybe it becomes merged :)

**Comment by KaKi87** [comment]
@Murmele Nice ! Would you please compile it, for 64-bit Windows ? Thanks

**Comment by Murmele** [comment]
At the moment it is quite unstable especially the discard function. I think the pipeline of the repository is creating a build, but I'm not sure. I'm not the developer of GitAhead

**Comment by FinlayDaG33k** [comment]
@Murmele It does create a build but it doesn't upload the artifacts (basically it discards the build)

**Comment by Murmele** [comment]
Ah ok I understand. Nevertheless the build fails for windows. At the moment I'm building only on windows

**Comment by KaKi87** [comment]
Hello,
Any news on this ?
Thanks

**Comment by KaKi87** [comment]
@nelsonroliveira What other products are you referring to ?

**Comment by nelsonroliveira** [comment]
gitkraken has it and if i remember correctly source tree also has it

**Comment by FinlayDaG33k** [comment]
@nelsonroliveira GitKraken indeed has this.
Too bad GitKraken removed the ability to work on private repos for free : <

**Comment by keverw** [comment]
Yes source tree has this.

I recently been using gitahead in a CentOS VM, and a better way to state things similar to source tree would be great. I wish source tree supported Linux, but at the same time Gitahead seems like a very close replacement even on the Mac.

**Comment by KaKi87** [comment]
Hello @hackhaslam

**Comment by zekiblue** [comment]
Any update for the feature?

**Comment by KaKi87** [comment]
Well, I eventually switched to JetBrains and I no longer need an external git client, I suggest you all do the same 😉

**Comment by FinlayDaG33k** [comment]
@KaKi87 I rather have this feature in GitAhead instead.
Switching to JetBrains will be quite cumbersome and expensive due to the sheer fact that I need a bunch of IDEs.
Also, I don't like JetBrains for the fact that I basically need a separate IDE for each language that I write in.
So no, I'll stick with my standard setup of VSCodium and GitAhead instead.

**Comment by KaKi87** [comment]
In that case, switch to SmartGit, which is freeware.

However, you don't need multiple IDEs when using IntelliJ, which supports everything that all the others combined do.

I didn't paid any JetBrains license yet, because they offer a student license which I've been using for 2 years now.

Once it expires, though, I will definitely pay for it, because it really increases my productivity, so it's worth the cost.

To finish, be aware that VSCodium is an editor, just like Sublime Text (which I also use for very light stuff, combined with SmartGit), not an IDE.

**Comment by FinlayDaG33k** [comment]
@KaKi87 SmartGit has a cluttered UI imo.
I have tried JetBrains in the past but wasn't very fond of it, especially not for the price they charge (I'm not a student and I don't apply for the OSL either).
So no, GitAhead + VSCodium it'll remain.

**Comment by KaKi87** [comment]
As you wish.
I had to move on from GitAhead, as this feature is **vital** to me, and we had no reply from the maintainer.

**Comment by safield** [comment]
You guys should learn to code. Then you can implement this feature yourself and submit a pull request.

**Comment by KaKi87** [comment]
I tried contributing to gitthermal and it lead anywhere, so I preferred switching to software that already implements the feature even if not FLOSS.

**Comment by KaKi87** [comment]
Plus, I always answer to requests and reports on my own projects.
Despite what the OSS license says, maintaining OSS *is* responsibility that should he taken seriously.

**Comment by Murmele** [comment]
The problem is that I don't have that much time to maintain a fork. I really like GitAhead and I'm working from time to time in my branch.

**Comment by FinlayDaG33k** [comment]
@safield That's pretty rude.
I think most people using GitAhead know how to code but might not use C++ to write their code.

**Comment by Philipp91** [comment]
While I'd be happy to contribute a feature like this if it's small, it doesn't seem like a small change in this case.

**Comment by Philipp91** [comment]
> but might not use C++ to write their code.

Right. Don't let that discourage you though!

While I'd be happy to contribute a feature like this if it's small, it doesn't seem to be a small change in this case. Quite a lot about the UX would have to change. Currently, GitAhead assumes that the state of a file can be shown in a _single_ view and then it just uses checkboxes to indicate whether a hunk is staged or not. When you stage a single line (through a different git interface, obviously), SmartGit will show a tri-state checkbox for the file, but it will falsely mark the whole hunk as staged. And you can't see which of the lines are actually staged. And if you think about it -- how would you? The green/red colors are already used for additions/removals. So the only way to visualize staged lines is to break them into individual hunks and have different checkboxes for each hunk. But that gets unwieldy and isn't user-friendly.

If you look at other Git UIs that do support line-based staging (e.g. Fork, or Atom's built-in Git plugin), you'll note that they have two separate file lists: One for staged files and one for unstaged files. When you (un)stage a whole file, it moves from one list to the other. However, if a file is partially staged (hunk-wise or line-wise), it appears in both lists. If you select it in the unstaged lists, it will shows all the diffs that aren't staged yet (offering you to stage them), and if you select it in the staged list, it shows the already-staged diffs (offering you to unstage them). So you can see two different sets of diffs on the same file.

Would the "product owner" of GitAhead consider changing the UI so much that there are no more checkboxes, but actually these two separate lists (inspired by these other Git tools)? Or can anyone propose an alternative UI that works well for partially staged hunks?

**Comment by Murmele** [comment]
@Philipp91 In my branch where I implemented line staging I added a marker next to the staged line and then the file checkbox gets partially selected. I will post an image tomorrow
The thing with two treeviews with staged and unstaged file I also implemented in my branch, it is quite comfortable

**Comment by Philipp91** [comment]
This looks great!

I tried building your fork, but it needs Qt 5.15, which isn't easily available from the Mint default package sources. Anyway, it would be great, if this feature could become part of the main release. @hackhaslam What do you think?

**Comment by Murmele** [comment]
I can try to make a flatpak. Or you can try and I can help you?

**Comment by Philipp91** [comment]
So you'd keep maintaining your fork without merging the changes here? Or is that just for testing?

I don't know how flatpak works in general. And I would first have to get Qt 5.15 running, I guess.

**Comment by FinlayDaG33k** [comment]
@Murmele That looks really neat.
Though instead of arrows, I'd just add a little "button" with a `+` or a `-` next to it.
Similar to this example from GitKraken (taken from their docs).
Something like that would make it feel less "foreign" imo.
![image](https://user-images.githubusercontent.com/13424273/118144774-1f70e900-b40d-11eb-8737-e54561fcb5a8.png)

**Comment by Murmele** [comment]
@FinlayDaG33k 
I'm using shortcuts  "s" and "u" for staging and unstaging. Is much faster than in gitkraken. Just select all line you would like to stage or unstage and press the shortcut. Even if you select non changed lines, they are just igored

**Comment by Philipp91** [comment]
While I don't want to request changes that take plenty of time to implement, I just want to point out the UX idea behind @FinlayDaG33k's proposal: If there's a "-" button next to something that unstages it, that inherently makes it obvious that it's currently staged. And vice versa. Much more so than an upward-pointing arrow.

**Comment by Murmele** [comment]
Yes I agree. I just took the first image I found in the editor, but perhaps at the moment I don't have time to do anything on Github

Don't know if it enough to change the url to my branch
https://github.com/flathub/io.github.gitahead.GitAhead/blob/master/io.github.gitahead.GitAhead.json

**Comment by Philipp91** [comment]
> I can try to make a flatpak. Or you can try and I can help you?

Would it be possible to set up some kind of Continuous Integration magic for this? I have very little experience with this (at least in the open-source world), but maybe it's possible for a GitHub Action to fire up some kind of minimal Debian that has Qt and all those nontrivial dependencies installed, run the build there, and then spit out the binary/flatpak/whatever? Ideally, I could attempt a code change locally and upload a PR, then it would build it there and give me the binary for download/testing.

**Comment by Murmele** [comment]
I think it is possible, there are already some CI Tests. They have to be fixed. I think it is also possible to create a dev version from te current master, but I don't know how

**Comment by SBCV** [comment]
Using `git gui` for this task might be an option, until such a feature has been integrated into `gitahead`

**Comment by mofobo-bit** [comment]
Must be a quick win, please !

**Comment by Murmele** [comment]
Please have a look into [Gittyup](https://murmele.github.io/Gittyup/) where you can stage single lines
